test/unittests/evm_test.cpp

```diff
@@ -153,6 +153,173 @@ TEST_P(evm, swapsn_push)
     EXPECT_STATUS(EVMC_BAD_JUMP_DESTINATION);
 }
 
+TEST_P(evm, dupn)
+{
+    // DUPN is not implemented in Advanced.
+    if (evm::isAdvanced())
+        return;
+
+    rev = EVMC_CANCUN;
+
+    auto pushes = bytecode{};
+    for (uint64_t i = 1; i <= 20; ++i)
+        pushes += push(i);
+
+    execute(pushes + OP_DUPN + "00" + ret_top());
+    EXPECT_STATUS(EVMC_SUCCESS);
+    EXPECT_OUTPUT_INT(20);
+
+    execute(pushes + OP_DUPN + "02" + ret_top());
+    EXPECT_STATUS(EVMC_SUCCESS);
+    EXPECT_OUTPUT_INT(18);
+
+    execute(pushes + OP_DUPN + "13" + ret_top());
+    EXPECT_STATUS(EVMC_SUCCESS);
+    EXPECT_OUTPUT_INT(1);
+
+    execute(pushes + OP_DUPN + "14" + ret_top());
+    EXPECT_STATUS(EVMC_STACK_UNDERFLOW);
+}
+
+TEST_P(evm, swapn)
+{
+    // SWAPN is not implemented in Advanced.
+    if (evm::isAdvanced())
+        return;
+
+    rev = EVMC_CANCUN;
+
+    auto pushes = bytecode{};
+    for (uint64_t i = 1; i <= 20; ++i)
+        pushes += push(i);
+
+    execute(pushes + OP_SWAPN + "00" + ret_top());
```

Comment:
You don't really check the other value that was swapped :smile: 
Maybe add another test with DUPN like
```cpp
    execute(pushes + OP_SWAPN + "00" + OP_DUPN + "01" + ret_top());
```

Reply:
Done.